R/Methods-FAKinGroupResults.R

```diff
@@ -217,85 +217,100 @@ setMethod("runSimulation", "FAKinGroupResults", function(object, nsim=50000,
-    BigRes <- lapply(Sims, FUN=function(x){
-        ExpRes <- do.call(rbind, lapply(affectedKins, function(z){
-            Res <- c(0, 0, 0)
-            expectedAffInKin <- intersect(x, z$phe)
-            Res[3] <- length(expectedAffInKin)
-            if(length(expectedAffInKin) >= length(z$aff)){
-                Res[1] <- 1
-            }
-            ## check frequency...
-            if(length(expectedAffInKin) >= 2){
-                ## that's the slowest part of the function... subsetting of the
-                ## kinship matrix...
-                ## we get some speedup if we subset a matrix, not a Matrix!
-                expectedAffInKin <- match(expectedAffInKin,
-                                          colnames(kinPhenotyped))
-                expectedKin <- kinPhenotyped[expectedAffInKin, expectedAffInKin]
-                diag(expectedKin) <- NA
-                ## get the table with the frequencies of kinship values...
-                expectedKinFreqs <- table(as.vector(expectedKin))
-                observedKinFreqs <- z$kinfreq
-                if(length(expectedKinFreqs) > 0 & length(observedKinFreqs) > 0){
-                    ## names of the table are the kinship values, ordered increasingly by size.
-                    expKinShipValue <- names(expectedKinFreqs[length(expectedKinFreqs)])
-                    obsKinShipValue <- names(observedKinFreqs[length(observedKinFreqs)])
-                    if(as.numeric(expKinShipValue) >= as.numeric(obsKinShipValue)){
-                        ## if the kinship value is the same, compare the counts.
-                        if(expKinShipValue==obsKinShipValue){
-                            if(expectedKinFreqs[expKinShipValue] >=
-                               observedKinFreqs[obsKinShipValue])
+    ExpRatio <- ExpKin <- rep(0, length(affectedKins))
+    expTable <- expDensity <- lapply(1:length(affectedKins), function(x) NULL )
+
+    nr.sim.todo <- nsim
+    ## Step size.
+    ## We compute small chunks of simulated data in order to keep the memory
+    ## footprint low. This involves strategies to add data to fixed vectors and
+    ## compute incremental tables and densities. A set of tables is ultimately
+    ## transformed to a histogram.
+
+    nr.inc.sim <- 1000
+    while( nr.sim.todo>0 ) {
+        ## Last simulation loop: run only the leftover number of steps.
+        if( nr.sim.todo<nr.inc.sim )
+            nr.inc.sim <- nr.sim.todo
+        if(is.null(strata)){
+            Sims <- lapply(1:nr.inc.sim, function(z){
+                return(sample(phenotypedIds, length(affectedIds)))
+            })
+        }else{
+            affStrataCounts <- table(strata[affectedIds])
+            affStrataCounts <- affStrataCounts[affStrataCounts > 0]
+            Sims <- lapply(1:nr.inc.sim, function(z){
+                idx <- stratsample(strata, counts=affStrataCounts)
+                return(phenotypedIds[idx])
+            })
+        }
+
+        BigRes <- list()
```

Comment:
Since we know the length of `BigRes` beforehand we could already create it with `BigRes <- vector("list", nr.inc.sim)`. We need than also to change line 290 below (next comment).